internal/crypto/crypto.go

```diff
@@ -233,91 +235,142 @@ func (k *EncryptionKey) Valid() bool {
 // holds the plaintext.
 var ErrInvalidCiphertext = errors.New("invalid ciphertext, same slice used for plaintext")
 
-// Encrypt encrypts and authenticates data. Stored in ciphertext is IV || Ciphertext ||
-// MAC. Encrypt returns the new ciphertext slice, which is extended when
-// necessary. ciphertext and plaintext may not point to (exactly) the same
-// slice or non-intersecting slices.
-func (k *Key) Encrypt(ciphertext []byte, plaintext []byte) ([]byte, error) {
+// validNonce checks that nonce is not all zero.
+func validNonce(nonce []byte) bool {
+	sum := 0
+	for b := range nonce {
+		sum += b
+	}
+	return sum > 0
+}
+
+// statically ensure that *Key implements crypto/cipher.AEAD
+var _ cipher.AEAD = &Key{}
+
+// NonceSize returns the size of the nonce that must be passed to Seal
+// and Open.
+func (k *Key) NonceSize() int {
+	return ivSize
+}
+
+// Overhead returns the maximum difference between the lengths of a
+// plaintext and its ciphertext.
+func (k *Key) Overhead() int {
+	return macSize
+}
+
+// Seal encrypts and authenticates plaintext, authenticates the
+// additional data and appends the result to dst, returning the updated
+// slice. The nonce must be NonceSize() bytes long and unique for all
+// time, for a given key.
+//
+// The plaintext and dst may alias exactly or not at all. To reuse
+// plaintext's storage for the encrypted output, use plaintext[:0] as dst.
+func (k *Key) Seal(dst, nonce, plaintext, additionalData []byte) []byte {
 	if !k.Valid() {
-		return nil, errors.New("invalid key")
+		panic("key is invalid")
 	}
 
-	ciphertext = ciphertext[:cap(ciphertext)]
+	if len(additionalData) > 0 {
+		panic("additional data is not supported")
+	}
 
-	// test for same slice, if possible
-	if len(plaintext) > 0 && len(ciphertext) > 0 && &plaintext[0] == &ciphertext[0] {
-		return nil, ErrInvalidCiphertext
+	if len(nonce) != ivSize {
+		panic("incorrect nonce length")
 	}
 
-	// extend ciphertext slice if necessary
-	if len(ciphertext) < len(plaintext)+Extension {
-		ext := len(plaintext) + Extension - len(ciphertext)
-		ciphertext = append(ciphertext, make([]byte, ext)...)
+	if !validNonce(nonce) {
+		panic("nonce is invalid")
 	}
 
-	iv := newIV()
-	copy(ciphertext, iv[:])
+	// extend dst so that the ciphertext fits
+	ciphertextLength := len(plaintext) + k.Overhead()
+	pos := len(dst)
+
+	capacity := cap(dst) - len(dst)
+	if capacity < ciphertextLength {
+		dst = dst[:cap(dst)]
+		dst = append(dst, make([]byte, ciphertextLength-capacity)...)
+	} else {
+		dst = dst[:pos+ciphertextLength]
+	}
```

Comment:
The common pattern here is `sliceForAppend`.

https://github.com/golang/go/blob/964639cc338db650ccadeafb7424bc8ebb2c0f6c/src/crypto/aes/aes_gcm.go#L81-L94

Reply:
Oh, I wasn't aware of that. Thanks for the pointer! The pattern is really nice and reduces the complexity of the code in Seal/Open :)